components/electric/lib/electric/satellite/client_reconnection_info.ex

```
@@ -260,41 +344,106 @@ defmodule Electric.Satellite.ClientReconnectionInfo do
         advance_graph_fn = Keyword.fetch!(opts, :advance_graph_using)
 
-        if CachedWal.Api.compare_positions(cached_wal_impl, acked_lsn, new_lsn) != :gt do
+        if CachedWal.Api.compare_positions(cached_wal_impl, acked_wal_pos, new_wal_pos) != :gt do
           received_data =
-            MapSet.new(txn_id_list, &{:transaction, &1})
-            |> MapSet.union(MapSet.new(subscription_id_list, &{:subscription, &1}))
-
-          new_graph =
-            graph
-            |> advance_up_to_new_lsn(
-              client_id,
-              cached_wal_impl,
-              origin,
-              advance_graph_fn,
-              acked_lsn,
-              new_lsn
+            MapSet.union(
+              MapSet.new(txids, &{:transaction, &1}),
+              MapSet.new(subscription_ids, &{:subscription, &1})
             )
-            |> advance_by_additional_data(client_id, received_data)
 
-          :ets.insert(@checkpoint_ets, {client_id, new_lsn, new_graph})
+          txn_stream =
+            CachedWal.Api.stream_transactions(cached_wal_impl, origin,
+              from: acked_wal_pos,
+              to: new_wal_pos
+            )
+
+          {new_graph, pending_actions, count, discarded_acc} =
+            advance_up_to_new_wal_pos(graph, advance_graph_fn, client_id, txn_stream)
+
+          Logger.debug(
+            "Advancing graph for #{inspect(client_id)} from #{inspect(acked_wal_pos)} to #{inspect(new_wal_pos)} by #{count} txns"
+          )
+
+          if map_size(pending_actions) > 0 do
+            :ets.insert(@actions_ets, Enum.to_list(pending_actions))
+          end
+
+          {new_graph, discarded_acc} =
+            advance_by_additional_data(new_graph, client_id, received_data, discarded_acc)
+
+          Client.with_pool(origin, fn ->
+            delete_discarded_cache_entries(discarded_acc)
+            store_client_checkpoint(client_id, new_wal_pos, new_graph)
+            store_client_actions(pending_actions)
```

nitpick: I'd appreciate a `!` in this function name, because my mind went immediately to the question of error handling in `with_pool/2` return value, but this uses a `query!` which will fail loudly instead - which I agree with, just want a pointer that this is what happens

---

Would you like to see all functions that call `Client.query!()`  inside to also have the ! in their name?

---

Yes, preferrably, just to have a reminder of how error handling is done here

---

If I do it transitively, almost every function in this module would get a `!` in its name.  I'm thinking of only applying this naming change to public functions, so that uses of this module's API in other modules communicate their error behaviour at a glance, while for private functions inside the module it can be seen from their implementation.

---

Just noting here that I've pushed a new commit that adds `!` to public function names only.